### PR TITLE
feat(java): support Lance Spark Batch Write

### PR DESCRIPTION
**Pull request by LuQQiu**
Add Lance Spark Batch Write Support.
Add a simple Lance catalog to support create table.
Batch write writes chunk to ArrowReader and zero-copy to Lance core.

TODO add the test cases for InternalRowWriterArrowReader
TODO larger scale manual testing

### CODE REVIEW

File: java/spark/src/main/java/com/lancedb/lance/spark/LanceCatalog.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.apache.spark.sql.connector.catalog.TableChange;
+import org.apache.spark.sql.connector.expressions.Transform;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+import scala.Some;
+
+import java.util.Map;
+
+public class LanceCatalog implements TableCatalog {
```
Comment (LuQQiu) [inline java/spark/src/main/java/com/lancedb/lance/spark/LanceCatalog.java:33]: Please review

File: java/spark/src/main/java/com/lancedb/lance/spark/write/InternalRowWriterArrowReader.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+import javax.annotation.concurrent.GuardedBy;
+import java.io.IOException;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A custom arrow reader that supports writes Spark internal rows while reading data in batches.
+ */
+public class InternalRowWriterArrowReader extends ArrowReader {
```
Comment (LuQQiu) [inline java/spark/src/main/java/com/lancedb/lance/spark/write/InternalRowWriterArrowReader.java:36]: Please review

File: java/spark/src/main/java/com/lancedb/lance/spark/write/LanceDataWriter.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import org.apache.spark.sql.connector.write.WriterCommitMessage;
+import org.apache.spark.sql.types.StructType;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.FutureTask;
+
+public class LanceDataWriter implements DataWriter<InternalRow> {
```
Comment (LuQQiu) [inline java/spark/src/main/java/com/lancedb/lance/spark/write/LanceDataWriter.java:32]: Please review

File: java/spark/src/test/java/com/lancedb/lance/spark/write/SparkWriteTest.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.apache.spark.sql.functions.col;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class SparkWriteTest {
```
Comment (LuQQiu) [inline java/spark/src/test/java/com/lancedb/lance/spark/write/SparkWriteTest.java:43]: Please review

File: java/spark/src/test/java/com/lancedb/lance/spark/write/InternalRowWriterArrowReaderTest.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      Thread writerThread = new Thread(() -> {
+        try {
+          for (int i = 0; i < totalRows; i++) {
+            InternalRow row = new GenericInternalRow(new Object[]{rowsWritten.incrementAndGet()});
+            arrowReader.write(row);
+            Thread.sleep(1);
+          }
+          arrowReader.setFinished();
+        } catch (Exception e) {
+          e.printStackTrace();
```
Comment (eddyxu) [inline java/spark/src/test/java/com/lancedb/lance/spark/write/InternalRowWriterArrowReaderTest.java:61]: This one swallows the exception?

File: java/spark/src/test/java/com/lancedb/lance/spark/write/LanceDataWriterTest.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class LanceDataWriterTest {
+  @TempDir
+  static Path tempDir;
+
+  @Test
+  public void testLanceDataWriter(TestInfo testInfo) throws IOException {
```
Comment (eddyxu) [inline java/spark/src/test/java/com/lancedb/lance/spark/write/LanceDataWriterTest.java:45]: This is Test only PR?

File: java/spark/src/main/java/com/lancedb/lance/spark/write/InternalRowWriterArrowReader.java (hunk 2 above)
Comment (eddyxu) [reply java/spark/src/main/java/com/lancedb/lance/spark/write/InternalRowWriterArrowReader.java:36]: Not clear what does this class do. `RowWriter` vs `ArrowReader`?

Can we have a better name? 
Is this used by `spark.write` or `spark.read`?

File: java/spark/src/main/java/com/lancedb/lance/spark/write/InternalRowWriterArrowReader.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    }
+    if (rowCount == 0) {
+      return false;
+    }
+    arrowWriter.finish();
+    // Calculate bytes read for the current record batch
+    // If the following code impacts performance, can be removed
+    VectorSchemaRoot root = this.getVectorSchemaRoot();
+    VectorUnloader unloader = new VectorUnloader(root);
+    try (ArrowRecordBatch recordBatch = unloader.getRecordBatch()) {
```
Comment (eddyxu) [inline java/spark/src/main/java/com/lancedb/lance/spark/write/InternalRowWriterArrowReader.java:112]: A `RecordBatch` is just consumed to update the `totalBytesRead`.

File: java/spark/src/main/java/com/lancedb/lance/spark/write/InternalRowWriterArrowReader.java (hunk 2 above)
Comment (eddyxu) [reply java/spark/src/main/java/com/lancedb/lance/spark/write/InternalRowWriterArrowReader.java:36]: Also can we use package public for this class?

File: java/spark/src/main/java/com/lancedb/lance/spark/write/LanceDataWriter.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    this.fragmentCreationTask = fragmentCreationTask;
+  }
+
+  @Override
+  public void write(InternalRow record) throws IOException {
+    arrowReader.write(record);
+  }
+
+  @Override
+  public WriterCommitMessage commit() throws IOException {
```
Comment (eddyxu) [inline java/spark/src/main/java/com/lancedb/lance/spark/write/LanceDataWriter.java:51]: This is commit at at the end of the full write? If so, we will accumulate too much data in `arrowReader`.

File: java/spark/src/main/java/com/lancedb/lance/spark/internal/LanceDatasetAdapter.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+public class LanceDatasetAdapter {
+  private static final BufferAllocator allocator = new RootAllocator(
+      RootAllocator.configBuilder().from(RootAllocator.defaultConfig())
+          .maxAllocation(4 * 1024 * 1024).build());
+
+  public static Optional<StructType> getSchema(LanceConfig config) {
+    return getSchema(config.getTablePath());
+  }
+
+  public static Optional<StructType> getSchema(String tablePath) {
```
Comment (eddyxu) [inline java/spark/src/main/java/com/lancedb/lance/spark/internal/LanceDatasetAdapter.java:48]: This does not look right. So we have `table` concept and `Dataset` concept here.  We need to be consistent. 

Also, can we just call this class `LanceDataset`?

File: java/spark/src/main/java/com/lancedb/lance/spark/internal/LanceDatasetAdapter.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+  public static FragmentMetadata createFragment(String datasetUri, ArrowReader reader) {
+    try (ArrowArrayStream arrowStream = ArrowArrayStream.allocateNew(allocator)) {
+      Data.exportArrayStream(allocator, reader, arrowStream);
+      return Fragment.create(datasetUri, arrowStream,
+          java.util.Optional.empty(), new WriteParams.Builder().build());
+    }
+  }
+
+  public static void createTable(String datasetUri, StructType sparkSchema) {
```
Comment (eddyxu) [inline java/spark/src/main/java/com/lancedb/lance/spark/internal/LanceDatasetAdapter.java:90]: be consistent of `dataset` or `table`

File: java/spark/src/main/java/com/lancedb/lance/spark/write/InternalRowWriterArrowReader.java (hunk 7 above)
Comment (LuQQiu) [reply java/spark/src/main/java/com/lancedb/lance/spark/write/InternalRowWriterArrowReader.java:112]: Yeah, there is no better way to compute bytes read. Alternative approach is to let bytesRead() throws unsupportedOperationException, in my testing, bytesRead is never called

File: java/spark/src/main/java/com/lancedb/lance/spark/write/LanceDataWriter.java (hunk 8 above)
Comment (LuQQiu) [reply java/spark/src/main/java/com/lancedb/lance/spark/write/LanceDataWriter.java:51]: The current sequence is
- create arrow reader
- start the reader thread and execute Fragment.create(reader)
- write to the arrow reader
- reader thread (fragment.create) consumes data in batches
- commit which notify there is no more new data and wait for reader thread (fragment.create) finishes

File: java/spark/src/main/java/com/lancedb/lance/spark/internal/LanceDatasetAdapter.java (hunk 10 above)
Comment (LuQQiu) [reply java/spark/src/main/java/com/lancedb/lance/spark/internal/LanceDatasetAdapter.java]: Changed to use `dataset`

File: java/spark/src/main/java/com/lancedb/lance/spark/internal/LanceDatasetAdapter.java (hunk 9 above)
Comment (LuQQiu) [reply java/spark/src/main/java/com/lancedb/lance/spark/internal/LanceDatasetAdapter.java]: this class is for having a central place to put allocator and put all methods requires allocator in the same space

The LanceTable has been renamed to LanceDataset which allows user to read or write a single lance dataset

**Comment by LuQQiu** [comment]
@eddyxu PTAL, thanks

**Review by eddyxu — APPROVED** [review]

**Comment by LuQQiu** [comment]
Merge this one for now.
Spark Write has the following main TODOs
1. commit fragments in once for all executor for one spark write job, instead of commit separately
2. Support FixedSizeList for Lance vector column

**Pull request merged**